fix: ppx and reason bug

Alizter opened the pull request:
Fix #7930

ppx needs the original source of a file even if it is a dialect. This was missing causing the sandboxed version to error. We add the original source file as a hidden dependency in the ppx action.

I also found that inline_tests was not using the correct dialect source, which can be seen in my commits as I tried to remove my original reason dependency. I therefore have also submitted a fix for that.

- [x] changelog

Alizter commented:
When a ppx action is sandboxed, it fails for reason files since for whatever reason, the ppx executable requires the original `hello.re` source when it really should only require `hello.re.ml` after we adapt for the dialect. The sandboxed action doesn't carry the original `.re` source hence the failure.

rgrinberg commented on src/dune_rules/preprocessing.ml:
```diff
@@ -703,13 +703,20 @@ let pp_one_module sctx ~lib_name ~scope ~preprocessor_deps
                         in
+                        let original_src : Dep.Set.t =
+                          match Module.source m ~ml_kind with
+                          | None -> Dep.Set.empty
```
How is `None` possible?

Alizter replied:
Not sure.

Alizter replied:
I guess I could code error here. I don't see any way this could be None.

rgrinberg replied:
It should be `Option.value_exn` then.

emillon commented:
I reproduced that without `reason` so I'll push a test which does not rely on a fake refmt.

emillon commented:
@Alizter I rebased and pushed a fix. Let me know what you think. This will need a changelog entry.  Thanks!

rgrinberg commented:
@emillon this PR needs review from me and @anmonteiro still

emillon commented:
sure, to be clear I was just addressing the test

Alizter commented:
@emillon It already has a changelog entry, is something missing?

emillon commented:
> @emillon It already has a changelog entry, is something missing?

ugh, coffee for me I guess. sorry!

anmonteiro commented on src/dune_rules/preprocessing.ml:
```diff
@@ -710,6 +710,10 @@ let pp_one_module sctx ~lib_name ~scope ~preprocessor_deps
                           ; Path (Path.build dst)
                           ; Command.Ml_kind.ppx_driver_flag ml_kind
                           ; Dep (Path.build src)
+                          ; Hidden_deps
```
I don't understand why this is needed. 

IIUC, the pipeline looks something like:

```
.re -> (dialect) .re.ml -> (pps) .pp.ml
```

So isn't the original source file already a (transitive) dependency of the pps action?

```

Alizter replied:
But sandboxing does not take the transitive closure, onoy the specified deoendencies. Thats why its important to always fully specify them.

Alizter commented:
@rgrinberg I've dropped the commit modifying the targets in inline_tests.ml. As you can see this creates an exception. I can fix it by tweaking the source files:
```diff
diff --git a/src/dune_rules/inline_tests.ml b/src/dune_rules/inline_tests.ml
index 77a92acc2..941900c09 100644
--- a/src/dune_rules/inline_tests.ml
+++ b/src/dune_rules/inline_tests.ml
@@ -285,7 +285,10 @@ include Sub_system.Register_end_point (struct
       in
       List.concat l
     in
-    let source_files = List.concat_map source_modules ~f:Module.sources in
+    let source_files =
+      List.concat_map source_modules ~f:(fun module_ ->
+          Module.ml_source module_ |> Module.sources)
+    in
     Memo.parallel_iter_set
       (module Mode_conf.Set)
       info.modes
diff --git a/test/blackbox-tests/test-cases/reason-expect.t/run.t b/test/blackbox-tests/test-cases/reason-expect.t/run.t
index d1c599f90..f694d3809 100644
--- a/test/blackbox-tests/test-cases/reason-expect.t/run.t
+++ b/test/blackbox-tests/test-cases/reason-expect.t/run.t
@@ -20,18 +20,7 @@ syntax and ppx_expect break at dune lang 3.3 due to the sandboxing of ppx.
   > (lang dune 3.9)
   > EOF
   $ dune test
-  File "dune", line 3, characters 1-15:
-  3 |  (inline_tests)
-       ^^^^^^^^^^^^^^
-  Fatal error: exception Sys_error("$TESTCASE_ROOT/_build/.sandbox/31552d8346d4edb1acb27f906d1a4b1b/default/./test.re.ml: No such file or directory")
-  Raised by primitive operation at Stdlib.open_in_gen in file "stdlib.ml", line 405, characters 28-54
-  Called from Stdio__In_channel.with_file in file "src/in_channel.ml", line 18, characters 45-66
-  Called from Stdio__In_channel.read_all in file "src/in_channel.ml" (inlined), line 86, characters 21-49
-  Called from Ppx_expect_evaluator.create_group in file "evaluator/ppx_expect_evaluator.ml", line 113, characters 22-69
-  Called from Base__List.count_map in file "src/list.ml", line 462, characters 13-17
-  Called from Ppx_expect_evaluator.evaluate_tests in file "evaluator/ppx_expect_evaluator.ml", line 207, characters 2-85
-  Called from Stdlib__List.map in file "list.ml", line 92, characters 20-23
-  Called from Stdlib__List.map in file "list.ml", line 92, characters 32-39
-  Called from Ppx_inline_test_lib__Runtime.exit in file "runtime-lib/runtime.ml", line 646, characters 2-49
-  Called from Dune__exe__Inline_test_runner_test in file ".test.inline-tests/inline_test_runner_test.ml-gen", line 1, characters 9-44
+  File "test.re.ml", line 1, characters 0-0:
+  Error: Files _build/default/test.re.ml and
+  _build/default/test.re.ml.corrected differ.
   [1]
```

Alizter commented:
@rgrinberg Shall I split the later commits into another PR?

rgrinberg commented on test/blackbox-tests/test-cases/reason-expect.t/refmt/refmt.ml:
```diff
@@ -0,0 +1,52 @@
+let sprintf = Printf.sprintf
```
Instead of copy pasting, what about moving this to `test-cases/utils` and making it available to both tests?

Alizter replied:
Its not an exact copy and paste which makes it awkward. One outputs to a file and the other outputs to stdout.

rgrinberg replied:
Shouldn't it just copy how the original refmt binary works? The fact that it behaves differently in different tests is wrong

rgrinberg commented:
I don't understand how this diff is supposed to work:

```
+  File "test.re.ml", line 1, characters 0-0:
+  Error: Files _build/default/test.re.ml and
+  _build/default/test.re.ml.corrected differ.
```

We have a source file `test.re`, and `test.re.ml` is supposed to contain the correction for a generated file?! Why would we offer a promotion for a generated file?

rgrinberg commented:
Okay, the issue is that our clone of refmt is bugged because it doesn't output line directives. Unfortunately, inserting the line directive to correct the filename breaks ppx_expect. No idea what's happening but I pushed the commit that demonstrates the issue.

Anyway, we can investigate this later. Just remove the copy-pasting between the refmt's and this is good to go

Alizter commented:
@rgrinberg Right, if you look at the commits, the last one I pushed fixes inline_tests. Before I get the same error as you did.

rgrinberg commented:
It's not the same error. The previous error was expected because the refmt error is broken. The subsequent is because ppx_expect doesn't understand directive for some reason

Alizter commented:
@rgrinberg I've consolidated the fake refmt and squashed your commit.

Alizter commented:
@rgrinberg Actually the backtrace in the bug you found isn't repro. My machine differs from the CI for some reason. Maybe we should leave the removing of the refmt dep to another PR and fix this inline tests issue?

rgrinberg commented:
We can't have the test suite depend on refmt

Alizter commented:
@rgrinberg Alright, shall I just scrub the backtrace then?

rgrinberg commented:
The back trace is easily fixable we just need to remove the locations.

That's not the point though, there's clearly something wrong with this fake refmt if it's making ppx_expect fail. We should fix it or it will break the test suite in other mysterious ways.

Alizter commented:
@rgrinberg You added this line to the fake refmt:
```
  output_string out (sprintf "# 1 %S\n" source);

```
That was intentional right?

rgrinberg commented:
Yes. Without this, how would ppx_expect find the original source?

anmonteiro commented:
What’s the refmt problem? Btw Reason can’t parse line directives, not sure if that’s related.

rgrinberg commented:
We have this fake `refmt` binary that we use for the tests. There's a problem with it's not working with ppx_expect for some reason (see exception). Perhaps you know what the reason is?

refmt doesn't need to understand line directives, it just needs to produce a parsetree with the correct locations. I think it does so with `--binary` which is what dune uses.

anmonteiro commented:
This looks like a ppx_expect bug rather than Reason? https://github.com/search?q=%22pos+%2B+len+past+end%22&type=code

anmonteiro commented:
My guess: the ppx_expect evaluator is trying to read the new file expecting the old file contents length?

Alizter commented:
Hmm I don't know how to progress with this now.
